# Objective-

1. Implemetaion of Web Scrapping using python
2. Sentence tokenization
3. How to improvise sentence tokenization
4. Word tokenization
5. Improvisation of word tokenisation
6. Conclusion

In [91]:
import sys
import requests
import bs4
import random
import re

### Function for wikipedia scrapping


In [92]:
def scrapeWikiArticle(url):
    response = requests.get(
        url=url
    )
    text=""
    soup = bs4.BeautifulSoup(response.content, 'html.parser')

    title = soup.find(id="firstHeading").getText()
    content = soup.find_all('p')
    text=text+title+" "
    
    for i in content:
        text=text+i.getText()
    return text

text=scrapeWikiArticle("https://en.wikipedia.org/wiki/COVID-19")


In [79]:
import nltk
from nltk.tokenize import sent_tokenize, word_tokenize

In [80]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ayush\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

### Sentence Tokenization

##### 1- Lets directly tokenize sentence and see what happens


In [96]:
#SENTENCES as tokens
sentences =sent_tokenize(text)

In [82]:
print("No of_sentences tokenized :", len(sentences))

No of_sentences tokenized : 480


###### we got total 480 sentences, lets see whats in there

In [97]:
for i in sentences:
    print(i)

COVID-19 
2020
2021
By conveyance
National responses
Vaccines
By country
Coronavirus disease 2019 (COVID-19) is a contagious disease caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2).
The first known case was identified in Wuhan, China, in December 2019.
[7] The disease has since spread worldwide, leading to an ongoing pandemic.
[8]
Symptoms of COVID-19 are variable, but often include fever,[9] cough, headache,[10] fatigue, breathing difficulties, and loss of smell and taste.
[11][12][13] Symptoms may begin one to fourteen days after exposure to the virus.
At least a third of people who are infected do not develop noticeable symptoms.
[14] Of those people who develop symptoms noticeable  enough to be classed as patients, most (81%) develop mild to moderate symptoms (up to mild pneumonia), while 14% develop severe symptoms (dyspnea, hypoxia, or more than 50% lung involvement on imaging), and 5% suffer critical symptoms (respiratory failure, shock, or multiorgan dysf

###### In above results u can see , we are getting setences with "[45], [67] etc", /n, so we got text but need to remove these unwanted text

In [102]:
# lets pre process the text and see how many sentences are there  after the correction
text_corr = re.sub(r'\[.*?\]+', '', text)
text_corr = sent_corr.replace('\n', '')

###### Lets see how many sentences got corrected

In [103]:
corr_sentences =sent_tokenize(text_corr)
print("No of_sentences tokenized :", len(corr_sentences))

No of_sentences tokenized : 345


In [104]:
# Total errors in found in sentence tokenization process
print("Total errors found: ",len(sentences)-len(corr_sentences))

Total errors found:  135


#### Reasons for error in sentence tokenization -

1. There were boxes like structure i.e [23],[25] etc
2. There were multiple /n
3. Sentence tokenizer was creating new sentence on every dot. So, if  it is Mr. Raman . it will break into two sentences.

### Word Tokenization

###### lets see directly tokenizing the words and see what it prints

In [105]:
words=word_tokenize(text)
print("No of words tokenized: ", len(words))

No of words tokenized:  14283


In [106]:
for i in words:
    print(i)

COVID-19
2020
2021
By
conveyance
National
responses
Vaccines
By
country
Coronavirus
disease
2019
(
COVID-19
)
is
a
contagious
disease
caused
by
severe
acute
respiratory
syndrome
coronavirus
2
(
SARS-CoV-2
)
.
The
first
known
case
was
identified
in
Wuhan
,
China
,
in
December
2019
.
[
7
]
The
disease
has
since
spread
worldwide
,
leading
to
an
ongoing
pandemic
.
[
8
]
Symptoms
of
COVID-19
are
variable
,
but
often
include
fever
,
[
9
]
cough
,
headache
,
[
10
]
fatigue
,
breathing
difficulties
,
and
loss
of
smell
and
taste
.
[
11
]
[
12
]
[
13
]
Symptoms
may
begin
one
to
fourteen
days
after
exposure
to
the
virus
.
At
least
a
third
of
people
who
are
infected
do
not
develop
noticeable
symptoms
.
[
14
]
Of
those
people
who
develop
symptoms
noticeable
enough
to
be
classed
as
patients
,
most
(
81
%
)
develop
mild
to
moderate
symptoms
(
up
to
mild
pneumonia
)
,
while
14
%
develop
severe
symptoms
(
dyspnea
,
hypoxia
,
or
more
than
50
%
lung
involvement
on
imaging
)
,
and
5
%
suffer
critical
symp

78
]
Cluster
5
emerged
among
minks
and
mink
farmers
in
Denmark
.
[
79
]
After
strict
quarantines
and
a
mink
euthanasia
campaign
,
the
cluster
was
assessed
to
no
longer
be
circulating
among
humans
in
Denmark
as
of
1
February
2021
.
[
80
]
As
of
July
2021
[
update
]
,
there
are
four
dominant
variants
of
SARS-CoV-2
spreading
among
global
populations
:
the
Alpha
Variant
(
formerly
called
the
UK
Variant
and
officially
referred
to
as
B.1.1.7
)
,
first
found
in
London
and
Kent
,
the
Beta
Variant
(
formerly
called
the
South
Africa
Variant
and
officially
referred
to
as
B.1.351
)
,
the
Gamma
Variant
(
formerly
called
the
Brazil
Variant
and
officially
referred
to
as
P.1
)
,
and
the
Delta
Variant
(
formerly
called
the
India
Variant
and
officially
referred
to
as
B.1.617.2
)
.
[
81
]
Using
whole
genome
sequencing
,
epidemiology
and
modelling
suggest
the
Alpha
variant
VUI-202012/01
(
the
first
variant
under
investigation
in
December
2020
)
in
the
B.1.1.7
lineage
transmits
more
easily
than
some
other


[
117
]
A
cytokine
storm
is
due
to
an
acute
hyperinflammatory
response
that
is
responsible
for
clinical
illness
in
an
array
of
diseases
but
in
COVID-19
,
it
is
related
to
worse
prognosis
and
increased
fatality
.
The
storm
causes
acute
respiratory
distress
syndrome
,
blood
clotting
events
such
as
strokes
,
myocardial
infarction
,
encephalitis
,
acute
kidney
injury
,
and
vasculitis
.
The
production
of
IL-1
,
IL-2
,
IL-6
,
TNF-alpha
,
and
interferon-gamma
,
all
crucial
components
of
normal
immune
responses
,
inadvertently
become
the
causes
of
a
cytokine
storm
.
The
cells
of
the
central
nervous
system
,
the
microglia
,
neurons
,
and
astrocytes
,
are
also
involved
in
the
release
of
pro-inflammatory
cytokines
affecting
the
nervous
system
,
and
effects
of
cytokine
storms
toward
the
CNS
are
not
uncommon
.
[
118
]
There
are
many
unknowns
for
pregnant
women
during
the
COVID-19
pandemic
.
Given
that
they
are
prone
to
suffering
from
complications
and
severe
disease
infection
with
other
types
of
co

[
239
]
[
240
]
A
majority
of
those
who
were
admitted
to
hospital
with
severe
disease
report
long-term
problems
including
fatigue
and
shortness
of
breath
.
[
241
]
On
30
October
2020
,
WHO
chief
Tedros
Adhanom
warned
that
``
to
a
significant
number
of
people
,
the
COVID
virus
poses
a
range
of
serious
long-term
effects
.
''
He
has
described
the
vast
spectrum
of
COVID-19
symptoms
that
fluctuate
over
time
as
``
really
concerning
''
.
They
range
from
fatigue
,
a
cough
and
shortness
of
breath
,
to
inflammation
and
injury
of
major
organs
–
including
the
lungs
and
heart
,
and
also
neurological
and
psychologic
effects
.
Symptoms
often
overlap
and
can
affect
any
system
in
the
body
.
Infected
people
have
reported
cyclical
bouts
of
fatigue
,
headaches
,
months
of
complete
exhaustion
,
mood
swings
,
and
other
symptoms
.
Tedros
therefore
concluded
that
a
strategy
of
achieving
herd
immunity
by
infection
,
rather
than
vaccination
,
is
``
morally
unconscionable
and
unfeasible
''
.
[
242
]
In
terms
of


prevalence
of
smoking
in
women
and
men
developing
co-morbid
conditions
such
as
hypertension
at
a
younger
age
than
women
could
have
contributed
to
the
higher
mortality
in
men
.
[
334
]
In
Europe
,
57
%
of
the
infected
people
were
men
and
72
%
of
those
died
with
COVID-19
were
men
.
[
335
]
As
of
April
2020
,
the
US
government
is
not
tracking
sex-related
data
of
COVID-19
infections
.
[
336
]
Research
has
shown
that
viral
illnesses
like
Ebola
,
HIV
,
influenza
and
SARS
affect
men
and
women
differently
.
[
336
]
In
the
US
,
a
greater
proportion
of
deaths
due
to
COVID-19
have
occurred
among
African
Americans
and
other
minority
groups
.
[
337
]
Structural
factors
that
prevent
them
from
practicing
social
distancing
include
their
concentration
in
crowded
substandard
housing
and
in
``
essential
''
occupations
such
as
retail
grocery
workers
,
public
transit
employees
,
health-care
workers
and
custodial
staff
.
Greater
prevalence
of
lacking
health
insurance
and
care
of
underlying
conditions
such
a

CDC
has
classified
it
to
require
a
BSL3
grade
laboratory
.
[
414
]
There
has
been
a
great
deal
of
COVID-19
research
,
involving
accelerated
research
processes
and
publishing
shortcuts
to
meet
the
global
demand
.
[
415
]
As
of
December
2020
[
update
]
,
hundreds
of
clinical
trials
have
been
undertaken
,
with
research
happening
on
every
continent
except
Antarctica
.
[
416
]
As
of
November
2020
[
update
]
,
more
than
200
possible
treatments
had
been
studied
in
humans
so
far
.
[
417
]
Modelling
research
has
been
conducted
with
several
objectives
,
including
predictions
of
the
dynamics
of
transmission
,
[
418
]
diagnosis
and
prognosis
of
infection
,
[
419
]
estimation
of
the
impact
of
interventions
,
[
420
]
[
421
]
or
allocation
of
resources
.
[
422
]
Modelling
studies
are
mostly
based
on
epidemiological
models
,
[
423
]
estimating
the
number
of
infected
people
over
time
under
given
conditions
.
Several
other
types
of
models
have
been
developed
and
used
during
the
COVID-19
including
comput

######  Now in above word tokenization, we could see, it is considering ".","5","]","[", everything a single words

In [119]:
#this step will remove the unwanted brackets as /n as words
corr_words=word_tokenize(text_corr)
print("No of words tokenized: ", len(corr_words))

#this step will remove the "(",")",",","." from the list
corr_words=list(filter(lambda a: a not in["(",")",".",",","``","''","'",":",";"], corr_words))
print("Final No of words tokenized AFTER all filtering: ", len(corr_words))

No of words tokenized:  12357
Final No of words tokenized AFTER all filtering:  11049


In [120]:
for i in corr_words:
    print(i)

COVID-19
20202021By
conveyanceNational
responsesVaccinesBy
countryCoronavirus
disease
2019
COVID-19
is
a
contagious
disease
caused
by
severe
acute
respiratory
syndrome
coronavirus
2
SARS-CoV-2
The
first
known
case
was
identified
in
Wuhan
China
in
December
2019
The
disease
has
since
spread
worldwide
leading
to
an
ongoing
pandemic.Symptoms
of
COVID-19
are
variable
but
often
include
fever
cough
headache
fatigue
breathing
difficulties
and
loss
of
smell
and
taste
Symptoms
may
begin
one
to
fourteen
days
after
exposure
to
the
virus
At
least
a
third
of
people
who
are
infected
do
not
develop
noticeable
symptoms
Of
those
people
who
develop
symptoms
noticeable
enough
to
be
classed
as
patients
most
81
%
develop
mild
to
moderate
symptoms
up
to
mild
pneumonia
while
14
%
develop
severe
symptoms
dyspnea
hypoxia
or
more
than
50
%
lung
involvement
on
imaging
and
5
%
suffer
critical
symptoms
respiratory
failure
shock
or
multiorgan
dysfunction
Older
people
are
at
a
higher
risk
of
developing
severe
symptom

–
changes
to
blood
cells
–
such
as
increased
sizes
–
sometimes
persisting
for
months
after
hospital
discharge.Another
common
cause
of
death
is
complications
related
to
the
kidneys
Early
reports
show
that
up
to
30
%
of
hospitalized
patients
both
in
China
and
in
New
York
have
experienced
some
injury
to
their
kidneys
including
some
persons
with
no
previous
kidney
problems.Autopsies
of
people
who
died
of
COVID-19
have
found
diffuse
alveolar
damage
and
lymphocyte-containing
inflammatory
infiltrates
within
the
lung.Although
SARS-CoV-2
has
a
tropism
for
ACE2-expressing
epithelial
cells
of
the
respiratory
tract
people
with
severe
COVID-19
have
symptoms
of
systemic
hyperinflammation
Clinical
laboratory
findings
of
elevated
IL‑2
IL‑7
IL‑6
granulocyte-macrophage
colony-stimulating
factor
GM‑CSF
interferon
gamma-induced
protein
10
IP‑10
monocyte
chemoattractant
protein
1
MCP1
macrophage
inflammatory
protein
1‑alpha
MIP‑1‑alpha
and
tumour
necrosis
factor
TNF‑α
indicative
of
cytokine
release
syndrom

surfaces
for
hours
to
days
If
a
person
touches
the
dirty
surface
they
may
deposit
the
virus
at
the
eyes
nose
or
mouth
where
it
can
enter
the
body
and
cause
infection
Evidence
indicates
that
contact
with
infected
surfaces
is
not
the
main
driver
of
COVID-19
leading
to
recommendations
for
optimised
disinfection
procedures
to
avoid
issues
such
as
the
increase
of
antimicrobial
resistance
through
the
use
of
inappropriate
cleaning
products
and
processes
Deep
cleaning
and
other
surface
sanitation
has
been
criticized
as
hygiene
theater
giving
a
false
sense
of
security
against
something
primarily
spread
through
the
air.The
amount
of
time
that
the
virus
can
survive
depends
significantly
on
the
type
of
surface
the
temperature
and
the
humidity
Coronaviruses
die
very
quickly
when
exposed
to
the
UV
light
in
sunlight
Like
other
enveloped
viruses
SARS-CoV-2
survives
longest
when
the
temperature
is
at
room
temperature
or
lower
and
when
the
relative
humidity
is
low
<
50
%
.On
many
surfaces
including
glas

ratio
or
infection
fatality
risk
This
metric
is
calculated
by
dividing
the
total
number
of
deaths
from
the
disease
by
the
total
number
of
infected
individuals
hence
in
contrast
to
the
CFR
the
IFR
incorporates
asymptomatic
and
undiagnosed
infections
as
well
as
reported
cases.A
December
2020
systematic
review
and
meta-analysis
estimated
that
population
IFR
during
the
first
wave
of
the
pandemic
was
about
0.5
%
to
1
%
in
many
locations
including
France
Netherlands
New
Zealand
and
Portugal
1
%
to
2
%
in
other
locations
Australia
England
Lithuania
and
Spain
and
exceeded
2
%
in
Italy
That
study
also
found
that
most
of
these
differences
in
IFR
reflected
corresponding
differences
in
the
age
composition
of
the
population
and
age-specific
infection
rates
in
particular
the
metaregression
estimate
of
IFR
is
very
low
for
children
and
younger
adults
e.g.
0.002
%
at
age
10
and
0.01
%
at
age
25
but
increases
progressively
to
0.4
%
at
age
55
1.4
%
at
age
65
4.6
%
at
age
75
and
15
%
at
age
85
These
resul

shortcuts
to
meet
the
global
demand.As
of
December
2020
hundreds
of
clinical
trials
have
been
undertaken
with
research
happening
on
every
continent
except
Antarctica
As
of
November
2020
more
than
200
possible
treatments
had
been
studied
in
humans
so
far.Modelling
research
has
been
conducted
with
several
objectives
including
predictions
of
the
dynamics
of
transmission
diagnosis
and
prognosis
of
infection
estimation
of
the
impact
of
interventions
or
allocation
of
resources
Modelling
studies
are
mostly
based
on
epidemiological
models
estimating
the
number
of
infected
people
over
time
under
given
conditions
Several
other
types
of
models
have
been
developed
and
used
during
the
COVID-19
including
computational
fluid
dynamics
models
to
study
the
flow
physics
of
COVID-19
retrofits
of
crowd
movement
models
to
study
occupant
exposure
mobility-data
based
models
to
investigate
transmission
or
the
use
of
macroeconomic
models
to
assess
the
economic
impact
of
the
pandemic
Further
conceptual
framework

###### Lets see the difference between ---- No of words after word tokenization AND No of words after filtering afterwards

In [121]:
print("Total errors in word tokenization: ", len(words)-len(corr_words))

Total errors in word tokenization:  1308


### Reasons of error-

1. All brackets [ , ] , ( ,) , commas , dots, ; , :  these all the things were treated as words 
2. /n were consider as words
3. if you directly copy the data from wikipedia , [] will be a word.

# Conclusion

We cannot directly apply tokenization on the web scrapped data. Due to different formattings it will give many more unnecessary outputs, which will lead to lesser efficiency of the project.

We need to pre-process the data before applying the tokenization.
Using regex we can filter various unnecessary segments from the text and can deliver the best possible text to tokenize, will result into maximum accuracy.